coffeescript-to-javascript-guidelines.md

```diff
@@ -0,0 +1,168 @@
+6. If this conversion is being done at the same time as a feature, submit a PR using the [interstitial PR process](#interstitial-pr-process)
+
+## Refactoring CoffeeScript magic
+### Classes & super
+#### CoffeeScript
+
+```coffee
+class CoreProduct extends Product
+
+  initialize: ->
+    super
+```
+
+#### Generated JavaScript
+This is what would be output from js2coffee
+
+```javascript
+var CoreProduct,
+  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
+  hasProp = {}.hasOwnProperty;
+
+CoreProduct = (function(superClass) {
+  extend(CoreProduct, superClass);
+
+  function CoreProduct() {
+    return CoreProduct.__super__.constructor.apply(this, arguments);
+  }
+
+  CoreProduct.prototype.initialize = function() {
+    return CoreProduct.__super__.initialize.apply(this, arguments);
+  };
+
+  return CoreProduct;
+
+})(Product);
+
+// ---
+// generated by coffee-script 1.9.2
+```
+
+#### Refactored JavaScript
+We can change from CoffeeScript's classes to Backbone's classes, like so
+
+```javascript
+var CoreProduct = Product.extend({
+  initialize: function() {
+    // we also change from using from `CoreProduct.__super__` to `Product.prototype`
+    return Product.prototype.initialize.apply(this, arguments);
+  }
+});
+```
+
+### Variables
+#### CoffeeScript
+
+```coffee
+_ = require('underscore')
+Product = require('models/product')
+TerminalMixins = require('presenters/mixins/terminal')
+```
+
+#### Generated JavaScript
+This is what would be output from js2coffee
+
+```javascript
+var Product, TerminalMixins, _;
+
+_ = require('underscore');
+
+Product = require('models/product');
+
+TerminalMixins = require('presenters/mixins/terminal');
+
+// ---
+// generated by coffee-script 1.9.2
+```
+
+#### Refactored JavaScript
+We would move the variable declarations to the same line that they are initialised, like so
+
+```javascript
+var _ = require('underscore');
+var Product = require('models/product');
+var TerminalMixins = require('presenters/mixins/terminal');
+```
+
+### Existence checks
+#### CoffeeScript
+
+```coffee
+_hasFlightCode: ->
+  @get('flights')?.out?.flight?.code
+```
+
+#### Generated JavaScript
+This is what would be output from js2coffee
+
+```javascript
+_hasFlightCode: function() {
+  var ref, ref1, ref2;
+  return (ref = this.get('flights')) != null ? (ref1 = ref.out) != null ? (ref2 = ref1.flight) != null ? ref2.code : void 0 : void 0 : void 0;
+}
+
+// ---
+// generated by coffee-script 1.9.2
+```
+
+#### Refactored JavaScript
+This is fairly unreadable, we would remove the temporary `ref` variables and do the following
+
+```javascript
+_hasFlightCode: function() {
+  var flights = this.get('flights');
+  if(!(flights && flights.out && flights.out.flight && flights.out.flight.code)) {
```

Comment:
This is probably personal preference but I find the original easier to follow? It tends to be the style I've seen when working on our JS projects. This is probably just an extreme example of a deeply nested property, hard to get around :(.